client/orm/db.go

```diff
@@ -819,63 +819,121 @@ func (d *dbBase) UpdateBatch(ctx context.Context, q dbQuerier, qs *querySet, mi
+	Q := d.ins.TableQuote()
+
+	buf := buffers.Get()
+	defer buffers.Put(buf)
+
+	_, _ = buf.WriteString("UPDATE ")
+	_, _ = buf.WriteString(Q)
+	_, _ = buf.WriteString(mi.Table)
+	_, _ = buf.WriteString(Q)
+
+	if d.ins.SupportUpdateJoin() {
+		_, _ = buf.WriteString(" T0 ")
+		_, _ = buf.WriteString(specifyIndexes)
+		_, _ = buf.WriteString(join)
+		_, _ = buf.WriteString(sets)
+		_, _ = buf.WriteString(" ")
+		_, _ = buf.WriteString(where)
+	} else {
+		_, _ = buf.WriteString(" ")
+		_, _ = buf.WriteString(sets)
+		_, _ = buf.WriteString(" WHERE ")
+		_, _ = buf.WriteString(Q)
+		_, _ = buf.WriteString(mi.Fields.Pk.Column)
+		_, _ = buf.WriteString(Q)
+		_, _ = buf.WriteString(" IN ( ")
+		_, _ = buf.WriteString("SELECT T0.")
+		_, _ = buf.WriteString(Q)
+		_, _ = buf.WriteString(mi.Fields.Pk.Column)
+		_, _ = buf.WriteString(Q)
+		_, _ = buf.WriteString(" FROM ")
+		_, _ = buf.WriteString(Q)
+		_, _ = buf.WriteString(mi.Table)
+		_, _ = buf.WriteString(Q)
+		_, _ = buf.WriteString(" T0 ")
+		_, _ = buf.WriteString(specifyIndexes)
+		_, _ = buf.WriteString(join)
+		_, _ = buf.WriteString(where)
+		_, _ = buf.WriteString(" )")
+	}
+
+	query := buf.String()
+
+	d.ins.ReplaceMarks(&query)
+
+	return query
+}
+
+func (d *dbBase) getSetSQL(cols []string, values []interface{}) string {
+
+	var T string
 
 	Q := d.ins.TableQuote()
 
 	if d.ins.SupportUpdateJoin() {
 		T = "T0."
 	}
 
-	cols := make([]string, 0, len(columns))
+	buf := buffers.Get()
+	defer buffers.Put(buf)
 
-	for i, v := range columns {
-		col := fmt.Sprintf("%s%s%s%s", T, Q, v, Q)
+	_, _ = buf.WriteString("SET ")
+
+	for i, v := range cols {
+		if i > 0 {
+			_, _ = buf.WriteString(", ")
+		}
+		_, _ = buf.WriteString(T)
+		_, _ = buf.WriteString(Q)
+		_, _ = buf.WriteString(v)
+		_, _ = buf.WriteString(Q)
+		_, _ = buf.WriteString(" = ")
 		if c, ok := values[i].(colValue); ok {
+			_, _ = buf.WriteString(T)
+			_, _ = buf.WriteString(Q)
+			_, _ = buf.WriteString(v)
+			_, _ = buf.WriteString(Q)
 			switch c.opt {
 			case ColAdd:
-				cols = append(cols, col+" = "+col+" + ?")
+				_, _ = buf.WriteString(" + ?")
 			case ColMinus:
-				cols = append(cols, col+" = "+col+" - ?")
+				_, _ = buf.WriteString(" - ?")
 			case ColMultiply:
-				cols = append(cols, col+" = "+col+" * ?")
+				_, _ = buf.WriteString(" * ?")
 			case ColExcept:
-				cols = append(cols, col+" = "+col+" / ?")
+				_, _ = buf.WriteString(" / ?")
 			case ColBitAnd:
-				cols = append(cols, col+" = "+col+" & ?")
+				_, _ = buf.WriteString(" & ?")
 			case ColBitRShift:
-				cols = append(cols, col+" = "+col+" >> ?")
+				_, _ = buf.WriteString(" >> ?")
 			case ColBitLShift:
-				cols = append(cols, col+" = "+col+" << ?")
+				_, _ = buf.WriteString(" << ?")
 			case ColBitXOR:
-				cols = append(cols, col+" = "+col+" ^ ?")
+				_, _ = buf.WriteString(" ^ ?")
 			case ColBitOr:
-				cols = append(cols, col+" = "+col+" | ?")
+				_, _ = buf.WriteString(" | ?")
```

Comment:
please add test cases to cover all these branches

Reply:
Okay.